lib/model/rwfolder.go

```diff
@@ -74,114 +74,108 @@ type dbUpdateJob struct {
-
-		folder:         cfg.ID,
-		dir:            cfg.Path(),
-		scanIntv:       time.Duration(cfg.RescanIntervalS) * time.Second,
-		ignorePerms:    cfg.IgnorePerms,
-		copiers:        cfg.Copiers,
-		pullers:        cfg.Pullers,
-		shortID:        shortID,
-		order:          cfg.Order,
-		maxConflicts:   cfg.MaxConflicts,
-		allowSparse:    !cfg.DisableSparseFiles,
-		checkFreeSpace: cfg.MinDiskFreePct != 0,
-
-		stop:        make(chan struct{}),
+		virtualMtimeRepo: db.NewVirtualMtimeRepo(model.db, cfg.ID),
+		dir:              cfg.Path(),
+		ignorePerms:      cfg.IgnorePerms,
+		copiers:          cfg.Copiers,
+		pullers:          cfg.Pullers,
+		order:            cfg.Order,
+		maxConflicts:     cfg.MaxConflicts,
+		allowSparse:      !cfg.DisableSparseFiles,
+		checkFreeSpace:   cfg.MinDiskFreePct != 0,
+
 		queue:       newJobQueue(),
 		pullTimer:   time.NewTimer(time.Second),
-		scanTimer:   time.NewTimer(time.Millisecond), // The first scan should be done immediately.
-		delayScan:   make(chan time.Duration),
-		scanNow:     make(chan rescanRequest),
 		remoteIndex: make(chan struct{}, 1), // This needs to be 1-buffered so that we queue a notification if we're busy doing a pull when it comes.
 
 		errorsMut: sync.NewMutex(),
 	}
 
-	if p.copiers == 0 {
-		p.copiers = defaultCopiers
+	f.configureCopiersAndPullers(cfg)
+
+	return f
+}
+
+func (f *rwFolder) configureCopiersAndPullers(config config.FolderConfiguration) {
+	if f.copiers == 0 {
+		f.copiers = defaultCopiers
 	}
-	if p.pullers == 0 {
-		p.pullers = defaultPullers
+	if f.pullers == 0 {
+		f.pullers = defaultPullers
 	}
 
-	if cfg.PullerPauseS == 0 {
-		p.pause = defaultPullerPause
+	if config.PullerPauseS == 0 {
+		f.pause = defaultPullerPause
 	} else {
-		p.pause = time.Duration(cfg.PullerPauseS) * time.Second
+		f.pause = time.Duration(config.PullerPauseS) * time.Second
 	}
 
-	if cfg.PullerSleepS == 0 {
-		p.sleep = defaultPullerSleep
+	if config.PullerSleepS == 0 {
+		f.sleep = defaultPullerSleep
 	} else {
-		p.sleep = time.Duration(cfg.PullerSleepS) * time.Second
+		f.sleep = time.Duration(config.PullerSleepS) * time.Second
 	}
-
-	return p
 }
 
 // Helper function to check whether either the ignorePerm flag has been
 // set on the local host or the FlagNoPermBits has been set on the file/dir
 // which is being pulled.
-func (p *rwFolder) ignorePermissions(file protocol.FileInfo) bool {
-	return p.ignorePerms || file.Flags&protocol.FlagNoPermBits != 0
+func (f *rwFolder) ignorePermissions(file protocol.FileInfo) bool {
+	return f.ignorePerms || file.Flags&protocol.FlagNoPermBits != 0
 }
 
 // Serve will run scans and pulls. It will return when Stop()ed or on a
 // critical error.
-func (p *rwFolder) Serve() {
-	l.Debugln(p, "starting")
-	defer l.Debugln(p, "exiting")
+func (f *rwFolder) Serve() {
+	l.Debugln(f, "starting")
+	defer l.Debugln(f, "exiting")
 
 	defer func() {
-		p.pullTimer.Stop()
-		p.scanTimer.Stop()
+		f.pullTimer.Stop()
+		f.scan.timer.Stop()
```

Comment:
Likewise, here we've broken out `scan`, but still reach in and manipulate it's attributes directly. We haven't gained much in that case. Better would be if the relevant stuff were wrapped in methods, making the `scan` a bit more opaque from the user's (`rwFolder`s) point of view.